openqa-advanced-retrigger-jobs

```diff
@@ -1,16 +1,88 @@
-done
+#!/usr/bin/env python3
+
+"""
+Retrigger openQA jobs based on database queries.
+
+Needs SSH access to the specified target openQA host.
+"""
+
+import argparse
+import logging
+import subprocess
+import sys
+from datetime import datetime
+
+logging.basicConfig()
+log = logging.getLogger(sys.argv[0] if __name__ == "__main__" else __name__)
+
+
+class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter, argparse.RawDescriptionHelpFormatter):
+    """Preserve multi-line __doc__ and provide default arguments in help strings."""
+
+    pass
+
+
+def main():
+    parser = argparse.ArgumentParser(description=__doc__, formatter_class=CustomFormatter)
+    parser.add_argument(
+        "-v",
+        "--verbose",
+        help="Increase verbosity level, specify multiple times to increase verbosity",
+        action="count",
+        default=1,
+    )
+    parser.add_argument("-H", "--host", default="openqa.opensuse.org", help="Target openQA host")
+    parser.add_argument(
+        "-s",
+        "--failed-since",
+        default=datetime.today().isoformat(),
+        help="Filter jobs failed since this date",
+    )
+    parser.add_argument("-w", "--worker", default=None, help="Filter jobs assigned to this worker")
+    parser.add_argument("-i", "--instance", default=None, help="Instance of the worker")
+    parser.add_argument("-r", "--result", default="incomplete", help="Filter jobs with this result")
+    parser.add_argument(
+        "-a",
+        "--additional-filters",
+        default=None,
+        help="Additional filters for the SQL query",
+    )
+    parser.add_argument("-c", "--comment", default=None, help="Comment to add to the retriggered jobs")
+    parser.add_argument(
+        "-d",
+        "--dry-run",
+        action="store_true",
+        help="If set, only print the actions without executing",
+    )
+    args = parser.parse_args()
+    verbose_to_log = {
+        0: logging.CRITICAL,
+        1: logging.ERROR,
+        2: logging.WARN,
+        3: logging.INFO,
+        4: logging.DEBUG,
+    }
+    logging_level = logging.DEBUG if args.verbose > 4 else verbose_to_log[args.verbose]
+    log.setLevel(logging_level)
+    instance_string = f" and instance='{args.instance}'" if args.instance else ""
+    worker_string = f"assigned_worker_id in (select id from workers where (host='{args.worker}'{instance_string})) and " if args.worker else ""
+    additional_filters = f" and {args.additional_filters}" if args.additional_filters else ""
+    client_prefix = "echo" if args.dry_run else ""
+
+    query = (
+        f"select id from jobs where ({worker_string}result='{args.result}' "
+        f"and clone_id is null and t_finished >= '{args.failed_since}'{additional_filters});"
+    )
+
+    log.debug(f"Using SQL query: '{query}' on {args.host}")
+    ssh_command = f'ssh {args.host} "sudo -u geekotest psql --no-align --tuples-only --command=\\"{query}\\" openqa"'
```

Comment:
Better use an array instead of a space separated string, otherwise a shell needs to parse it.

Reply:
done